In [1]:
from yelpapi import YelpAPI
import pandas as pd
import json
import os, json, math, time
with open('/Users/dmusl/.secret/yelp_api.json') as f:
    login = json.load(f)
login.keys()

dict_keys(['client-id', 'api-key'])

In [2]:
yelp_api = YelpAPI(login['api-key'], timeout_s=5.0)
yelp_api

In [3]:
# set our API call parameters and filename before the first call
location = 'North Olmsted, OH 44070'
term = 'plant-based'

In [4]:
location.split(',')[0]

'North Olmsted'

In [5]:
## Specify folder for saving data
FOLDER = 'Data/'

os.makedirs(FOLDER, exist_ok = True)
# Specifying JSON_FILE filename (can include a folder)
JSON_FILE = FOLDER+f"{location.split(',')[0]}-{term}.json"

In [6]:
# use our yelp_api variable's search_query method to perform our API call
results = yelp_api.search_query(term = term, location = location)

In [7]:
type(results)
results.keys()

dict_keys(['businesses', 'total', 'region'])

In [8]:
results['total']

33

In [9]:
## how many businesses in our results
len(results['businesses'])

20

In [10]:
pb_f = pd.DataFrame(results['businesses'])
pb_f.head() 

id                                              alias  \
0  wv6LOnEpveAY37yGRHDAzQ                bibibop-asian-grill-fairview-park-3   
1  7hrC58k9Opx_B1HX__e4pg                         propaganda-coffee-lakewood   
2  KEAfojRAlmhICSUYWsHZ_Q                      mcalisters-deli-north-olmsted   
3  VXVrYjLRqNpQY-ogN9ztmQ                           cleveland-vegan-lakewood   
4  o_6W4vPRdG8eO7z2SIc0_Q  robeks-fresh-juices-and-smoothies-north-olmsted-4   

                              name  \
0              BIBIBOP Asian Grill   
1                Propaganda Coffee   
2                 McAlister's Deli   
3                  Cleveland Vegan   
4  Robeks Fresh Juices & Smoothies   

                                           image_url  is_closed  \
0  https://s3-media4.fl.yelpcdn.com/bphoto/5EUJjY...      False   
1  https://s3-media4.fl.yelpcdn.com/bphoto/PE7oqu...      False   
2  https://s3-media4.fl.yelpcdn.com/bphoto/0wpGH1...      False   
3  https://s3-media4.fl.yelpcdn.com/bphoto/-Do3Co...      False   
4  https://s3-media2.fl.yelpcdn.com/bphoto/blpuhN...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/bibibop-asian-grill-f...            21   
1  https://www.yelp.com/biz/propaganda-coffee-lak...            27   
2  https://www.yelp.com/biz/mcalisters-deli-north...            54   
3  https://www.yelp.com/biz/cleveland-vegan-lakew...           296   
4  https://www.yelp.com/biz/robeks-fresh-juices-a...            22   

                                          categories  rating  \
0  [{'alias': 'salad', 'title': 'Salad'}, {'alias...     4.0   
1  [{'alias': 'vegan', 'title': 'Vegan'}, {'alias...     4.5   
2  [{'alias': 'salad', 'title': 'Salad'}, {'alias...     4.5   
3  [{'alias': 'vegan', 'title': 'Vegan'}, {'alias...     4.5   
4  [{'alias': 'juicebars', 'title': 'Juice Bars &...     3.5   

                                         coordinates        transactions  \
0  {'latitude': 41.4615742, 'longitude': -81.8497...  [pickup, delivery]   
1     {'latitude': 41.48474, 'longitude': -81.82057}          [delivery]   
2  {'latitude': 41.41740677707591, 'longitude': -...          [delivery]   
3  {'latitude': 41.4855267595181, 'longitude': -8...          [delivery]   
4  {'latitude': 41.43469705946628, 'longitude': -...  [pickup, delivery]   

                                            location         phone  \
0  {'address1': '3091 Westgate Mall', 'address2':...  +14405531599   
1  {'address1': '17806 Detroit Ave', 'address2': ...  +12167675811   
2  {'address1': '356 Great Northern Mall', 'addre...  +14407344440   
3  {'address1': '17112 Detroit Ave', 'address2': ...  +12162210201   
4  {'address1': '23420 Lorain Rd', 'address2': 'S...  +14408011610   

    display_phone      distance price  
0  (440) 553-1599   7049.063265   NaN  
1  (216) 767-5811  10575.475468   NaN  
2  (440) 734-4440    762.235430    $$  
3  (216) 221-0201  10908.787939    $$  
4  (440) 801-1610   3147.589167    $$

In [11]:
# use our yelp_api variable's search_query method to perform our API call
results = yelp_api.search_query(term = term, location = location, offset = 20)

In [12]:
pb_f2 = pd.DataFrame(results['businesses'])
pb_f2.head() 

id                       alias         name  \
0  NDKBMYjNhckqg_lf53plug   burger-king-fairview-park  Burger King   
1  P6A79lfKj76LbJv9j3rDzg  burger-king-strongsville-2  Burger King   
2  mzGrMnuHJpPJEF7E4esGqw     burger-king-rocky-river  Burger King   
3  _E5sWHHe0tlVPuU6pR1gVg    burger-king-cleveland-37  Burger King   
4  XUcJiJx4x1RBtE31XrNexw          market-rocky-river       Market   

                                           image_url  is_closed  \
0  https://s3-media1.fl.yelpcdn.com/bphoto/sUDQHM...      False   
1  https://s3-media1.fl.yelpcdn.com/bphoto/1kq9Uu...      False   
2  https://s3-media3.fl.yelpcdn.com/bphoto/H_T5Lh...      False   
3  https://s3-media4.fl.yelpcdn.com/bphoto/eXm4z4...      False   
4  https://s3-media2.fl.yelpcdn.com/bphoto/VZsrcl...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/burger-king-fairview-...            13   
1  https://www.yelp.com/biz/burger-king-strongsvi...             6   
2  https://www.yelp.com/biz/burger-king-rocky-riv...            15   
3  https://www.yelp.com/biz/burger-king-cleveland...             8   
4  https://www.yelp.com/biz/market-rocky-river?ad...           266   

                                          categories  rating  \
0  [{'alias': 'burgers', 'title': 'Burgers'}, {'a...     1.5   
1  [{'alias': 'burgers', 'title': 'Burgers'}, {'a...     2.5   
2  [{'alias': 'burgers', 'title': 'Burgers'}, {'a...     1.5   
3  [{'alias': 'burgers', 'title': 'Burgers'}, {'a...     2.0   
4  [{'alias': 'breakfast_brunch', 'title': 'Break...     3.5   

                                         coordinates transactions price  \
0  {'latitude': 41.4387673237211, 'longitude': -8...   [delivery]     $   
1   {'latitude': 41.348393, 'longitude': -81.823185}   [delivery]   NaN   
2  {'latitude': 41.4637350475144, 'longitude': -8...   [delivery]     $   
3   {'latitude': 41.419197, 'longitude': -81.832565}   [delivery]     $   
4   {'latitude': 41.481787, 'longitude': -81.840207}   [delivery]    $$   

                                            location         phone  \
0  {'address1': '22740 Lorain Rd', 'address2': ''...  +14407342200   
1  {'address1': '8300 Pearl Rd', 'address2': None...  +14403842162   
2  {'address1': '20333 Center Ridge Rd', 'address...  +14403319279   
3  {'address1': '19000 Brookpark Rd', 'address2':...  +12164659551   
4  {'address1': '1137 Linda St', 'address2': '', ...  +14407994292   

    display_phone      distance  
0  (440) 734-2200   4852.393081  
1  (440) 384-2162  11210.324357  
2  (440) 331-9279   8303.740522  
3  (216) 465-9551   7488.982920  
4  (440) 799-4292   9979.202636

In [13]:
## concatenate the previous results and new results. 
pb = pd.concat([pb_f, pb_f2],
                      ignore_index=True)
display(pb.head(), pb.tail())

id                                              alias  \
0  wv6LOnEpveAY37yGRHDAzQ                bibibop-asian-grill-fairview-park-3   
1  7hrC58k9Opx_B1HX__e4pg                         propaganda-coffee-lakewood   
2  KEAfojRAlmhICSUYWsHZ_Q                      mcalisters-deli-north-olmsted   
3  VXVrYjLRqNpQY-ogN9ztmQ                           cleveland-vegan-lakewood   
4  o_6W4vPRdG8eO7z2SIc0_Q  robeks-fresh-juices-and-smoothies-north-olmsted-4   

                              name  \
0              BIBIBOP Asian Grill   
1                Propaganda Coffee   
2                 McAlister's Deli   
3                  Cleveland Vegan   
4  Robeks Fresh Juices & Smoothies   

                                           image_url  is_closed  \
0  https://s3-media4.fl.yelpcdn.com/bphoto/5EUJjY...      False   
1  https://s3-media4.fl.yelpcdn.com/bphoto/PE7oqu...      False   
2  https://s3-media4.fl.yelpcdn.com/bphoto/0wpGH1...      False   
3  https://s3-media4.fl.yelpcdn.com/bphoto/-Do3Co...      False   
4  https://s3-media2.fl.yelpcdn.com/bphoto/blpuhN...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/bibibop-asian-grill-f...            21   
1  https://www.yelp.com/biz/propaganda-coffee-lak...            27   
2  https://www.yelp.com/biz/mcalisters-deli-north...            54   
3  https://www.yelp.com/biz/cleveland-vegan-lakew...           296   
4  https://www.yelp.com/biz/robeks-fresh-juices-a...            22   

                                          categories  rating  \
0  [{'alias': 'salad', 'title': 'Salad'}, {'alias...     4.0   
1  [{'alias': 'vegan', 'title': 'Vegan'}, {'alias...     4.5   
2  [{'alias': 'salad', 'title': 'Salad'}, {'alias...     4.5   
3  [{'alias': 'vegan', 'title': 'Vegan'}, {'alias...     4.5   
4  [{'alias': 'juicebars', 'title': 'Juice Bars &...     3.5   

                                         coordinates        transactions  \
0  {'latitude': 41.4615742, 'longitude': -81.8497...  [pickup, delivery]   
1     {'latitude': 41.48474, 'longitude': -81.82057}          [delivery]   
2  {'latitude': 41.41740677707591, 'longitude': -...          [delivery]   
3  {'latitude': 41.4855267595181, 'longitude': -8...          [delivery]   
4  {'latitude': 41.43469705946628, 'longitude': -...  [pickup, delivery]   

                                            location         phone  \
0  {'address1': '3091 Westgate Mall', 'address2':...  +14405531599   
1  {'address1': '17806 Detroit Ave', 'address2': ...  +12167675811   
2  {'address1': '356 Great Northern Mall', 'addre...  +14407344440   
3  {'address1': '17112 Detroit Ave', 'address2': ...  +12162210201   
4  {'address1': '23420 Lorain Rd', 'address2': 'S...  +14408011610   

    display_phone      distance price  
0  (440) 553-1599   7049.063265   NaN  
1  (216) 767-5811  10575.475468   NaN  
2  (440) 734-4440    762.235430    $$  
3  (216) 221-0201  10908.787939    $$  
4  (440) 801-1610   3147.589167    $$

id                                       alias  \
26  6dDrLNa3_fx9E4lfKHtuig                        mcdonalds-westlake-2   
27  WBay6AZsHMExtMQsRc9-bQ     mellow-mushroom-rocky-river-rocky-river   
28  wNMUtWy0SRQu7m3i4FNhYg  red-robin-gourmet-burgers-and-brews-avon-2   
29  _NzimzgKxXdfnCv9skj71A                    donatos-pizza-lakewood-5   
30  86Rap6uerp9zK3y5xlhARw                                   zoup-avon   

                                   name  \
26                           McDonald's   
27          Mellow Mushroom Rocky River   
28  Red Robin Gourmet Burgers and Brews   
29                        Donatos Pizza   
30                                Zoup!   

                                            image_url  is_closed  \
26  https://s3-media1.fl.yelpcdn.com/bphoto/-lpQrx...      False   
27  https://s3-media4.fl.yelpcdn.com/bphoto/D8_8EO...      False   
28  https://s3-media2.fl.yelpcdn.com/bphoto/k8p6aZ...      False   
29  https://s3-media2.fl.yelpcdn.com/bphoto/tNspjB...      False   
30  https://s3-media3.fl.yelpcdn.com/bphoto/44P7AN...      False   

                                                  url  review_count  \
26  https://www.yelp.com/biz/mcdonalds-westlake-2?...            28   
27  https://www.yelp.com/biz/mellow-mushroom-rocky...           329   
28  https://www.yelp.com/biz/red-robin-gourmet-bur...            88   
29  https://www.yelp.com/biz/donatos-pizza-lakewoo...            47   
30  https://www.yelp.com/biz/zoup-avon?adjust_crea...            31   

                                           categories  rating  \
26  [{'alias': 'hotdogs', 'title': 'Fast Food'}, {...     2.0   
27  [{'alias': 'pizza', 'title': 'Pizza'}, {'alias...     3.5   
28  [{'alias': 'burgers', 'title': 'Burgers'}, {'a...     3.5   
29  [{'alias': 'pizza', 'title': 'Pizza'}, {'alias...     2.0   
30  [{'alias': 'salad', 'title': 'Salad'}, {'alias...     3.5   

                                          coordinates        transactions  \
26  {'latitude': 41.4704912756971, 'longitude': -8...          [delivery]   
27  {'latitude': 41.4813361, 'longitude': -81.836549}  [delivery, pickup]   
28     {'latitude': 41.46115, 'longitude': -82.01283}  [delivery, pickup]   
29     {'latitude': 41.48447, 'longitude': -81.82308}  [delivery, pickup]   
30  {'latitude': 41.46068986, 'longitude': -82.015...  [delivery, pickup]   

                                             location         phone  \
26  {'address1': '25700 Detroit Rd', 'address2': N...  +14408719161   
27  {'address1': '19333 Detroit Rd', 'address2': N...  +14403567171   
28  {'address1': '35858 Detroit Rd', 'address2': '...  +14409370260   
29  {'address1': '18100 Detroit Ave', 'address2': ...  +12162277200   
30  {'address1': '35918 Detroit Rd', 'address2': '...  +14406953939   

     display_phone      distance price  
26  (440) 871-9161   6200.434402     $  
27  (440) 356-7171  10147.741883    $$  
28  (440) 937-0260   9035.563957    $$  
29  (216) 227-7200  11203.844118    $$  
30  (440) 695-3939   9184.105523     $

In [14]:
# save the results to records-style json file with appropriate name
pb.to_json("Data/yelp-search-plant_based-north_olmsted.json", orient='records')


In [37]:
## Check if JSON_FILE exists
file_exists = os.path.isfile(JSON_FILE)
## If it does not exist: 
if file_exists == False:
    
    ## CREATE ANY NEEDED FOLDERS
    # Get the Folder Name only
    folder = os.path.dirname(JSON_FILE)
    ## If JSON_FILE included a folder:
    if len(folder)>0:
        # create the folder
        os.makedirs(folder,exist_ok=True)
        
        
    ## INFORM USER AND SAVE EMPTY LIST
    print(f'[i] {JSON_FILE} not found. Saving empty list to file.')
    
    
    # save an empty list
    with open(JSON_FILE,'w') as f:
        json.dump([],f)  
# If it exists, inform user
else:
    print(f"[i] {JSON_FILE} already exists.")

[i] Data/results_in_progress_NO_vegan.json not found. Saving empty list to file.


# Determine how many results are already in the file

Load the results file to determine the # of results we have previously retrieved. If you just created the file, you would expect it to be empty.

We will use this as our offset parameter for our API call. Even if this is your first API call, and the number is 0, we want to define "n_results" based on the length of "previous_results."

In [38]:
## Load previous results and use len of results for offset
with open(JSON_FILE,'r') as f:
    previous_results = json.load(f)
    
## set offset based on previous results
n_results = len(previous_results)
print(f'- {n_results} previous results found.')

- 0 previous results found.


# Figure out how many pages of results we will need

We will perform our first query to get our first page of results and the total number of results. We will then (via code) calculate how many pages we will need to retrieve all of our results.

In [40]:
# use our yelp_api variable's search_query method to perform our API call
results = yelp_api.search_query(term = term, location = location, offset = n_results)

results.keys()

dict_keys(['businesses', 'total', 'region'])

In [41]:
## How many results total?
total_results = results['total']
total_results

180

In [42]:
## How many did we get the details for?
results_per_page = len(results['businesses'])
results_per_page

20

There are over 11000 businesses to retrieve from our API, and we can get 20 results at a time (per "page").
We can calculate the # of results remaining by subtracting our offset (length of our previous results) from our total.
Then we can determine how many pages we will need by dividing the results by 20 (or whatever the value happens to be for results per page)
Note that we need to round up the number of pages in order to get all of the results. Even if there is only 1 result on the last page, we want to include that page! To do this, we will use math.ceil.

In [43]:
# Import additional packages for controlling our loop
import time, math
# Use math.ceil to round up for the total number of pages of results.
n_pages = math.ceil((results['total']-n_results)/ results_per_page)
n_pages

9

When this example was written, there were 11200 results and 20 results per page. 180 /20 = 9. But what if we had 11210 results? Rounding UP would give us 561 pages. We expect the first 560 pages to have 20 results and the last page to have the final 10 results. Notice that we have assigned the number of pages as n_pages here. We will use this value in our next segment of code.

You can see that having to manually go through 560 pages would be quite time-consuming and inefficient! First, we are going to save the first page into our file, and then we will add to it with a for loop.

In [44]:
# join new results with old list with extend and save to file
previous_results.extend(results['businesses'])  
with open(JSON_FILE,'w') as f:
     json.dump(previous_results,f)

# Set up a progress bar in our for loop.

To keep us informed about where we are in our loop, we will add a progress bar to our for loop.

TQDM is a package designed for adding animated progress bars to Python processes. It is not currently included in your dojo-env, so you are going to install it manually by opening a new Terminal/GitBash window and running the following command:

TQDM is easy to use and simply needs to know what we are looping through. If you wanted to test tqdm in action, but your loop is too fast, you can import time and use time.sleep to add a pause within your for loop. We will also use time.sleep when executing many API calls so that we do not overwhelm the server.

In [45]:
from tqdm.notebook import tqdm_notebook
import time
for i in tqdm_notebook(range(n_pages)):
    # adds 200 ms pause
    time.sleep(.2) 

  0%|          | 0/9 [00:00<?, ?it/s]

# For Loop to call each page
The loop below will iterate through each page of the results by starting at the appropriate offset. It will then append the results to the previous_results. This may take some time, so check out the progress bar!

In [46]:
for i in tqdm_notebook( range(1,n_pages+1)):
    
    ## Read in results in progress file and check the length
    with open(JSON_FILE, 'r') as f:
        previous_results = json.load(f)
    ## save number of results for to use as offset
    n_results = len(previous_results)
    ## use n_results as the OFFSET 
    results = yelp_api.search_query(location=location,
                                    term=term, 
                                    offset=n_results)
    
    ## append new results and save to file
    previous_results.extend(results['businesses'])
    
    with open(JSON_FILE,'w') as f:
        json.dump(previous_results,f)
    
    # add a 200ms pause
    time.sleep(.2)

  0%|          | 0/9 [00:00<?, ?it/s]

# Handling queries with >1000 Results
To get around this error, we can add an extra logic check to see if the length of movies we have so far (n_results) + the # of results on each page (results_per_page) is greater than 1,000.

If it is greater than 1,000, we will use a break to end our loop early.

# Deleting Our Previous Results FIle
Let's give ourselves a fresh start with our new and improved loop. Let's delete our previous results file.
We could accomplish this manually or programmatically## .
The Manual Way
We could do this manually by using the Files page in Jupyter to find the file in our results folder and delete it. To delete a file with Jupyter check the check box next to the file you want to delete.

You should now see additional buttons appear above the list of files.

Click on the red trash can icon to delete the file.



# The Programmatic Way
Just like we used the os module to create new folders, we can also use it to delete (or remove) files.

We simply use the os.remove function and pass in our filename to delete.

We can then use our os.path.isfile function again to confirm the file no longer exists.

In [47]:
## delete file and confirm it no longer exits.
os.remove(JSON_FILE)
os.path.isfile(JSON_FILE)

False

Now that we've deleted the file, we need to re-run our code to create it.

This process is begging to be turned into a function so we can easily repeat it.

While we are making it a function, we will add the option to delete the JSON file if it already exists, just like we did above.

So let's make a create_json_file function that accepts the JSON_FILE filename as the first argument and a second argument called delete_if_exists and set it to False by default.

This way, it will not automatically delete previous search results. We will have to explicitly say delete_if_exists = True to do so.

In [49]:
def create_json_file(JSON_FILE,  delete_if_exists=False):
    
    ## Check if JSON_FILE exists
    file_exists = os.path.isfile(JSON_FILE)
    
    ## If it DOES exist:
    if file_exists == True:
        
        ## Check if user wants to delete if exists
        if delete_if_exists==True:
            
            print(f"[!] {JSON_FILE} already exists. Deleting previous file...")
            ## delete file and confirm it no longer exits.
            os.remove(JSON_FILE)
            ## Recursive call to function after old file deleted
            create_json_file(JSON_FILE,delete_if_exists=False)
        else:
            print(f"[i] {JSON_FILE} already exists.")            
            
            
    ## If it does NOT exist:
    else:
        
        ## INFORM USER AND SAVE EMPTY LIST
        print(f"[i] {JSON_FILE} not found. Saving empty list to new file.")
        
        ## CREATE ANY NEEDED FOLDERS
        # Get the Folder Name only
        folder = os.path.dirname(JSON_FILE)
        
        ## If JSON_FILE included a folder:
        if len(folder)>0:
            # create the folder
            os.makedirs(folder,exist_ok=True)
        ## Save empty list to start the json file
        with open(JSON_FILE,'w') as f:
            json.dump([],f)  

Now that we have our new function, we can use it with delete_if_exists=True to delete our previous results and start over. We will also need to repeat the steps to recreate our n-results, total_results, results_per_page, and n_pages variables that we created before our first attempted loop.

In [50]:
## Create a new empty json file (exist the previous if it exists)
create_json_file(JSON_FILE, delete_if_exists=True)
## Load previous results and use len of results for offset
with open(JSON_FILE,'r') as f:
    previous_results = json.load(f)
    
## set offset based on previous results
n_results = len(previous_results)
print(f'- {n_results} previous results found.')
# use our yelp_api variable's search_query method to perform our API call
results = yelp_api.search_query(location=location,
                                term=term,
                               offset=n_results)
## How many results total?
total_results = results['total']
## How many did we get the details for?
results_per_page = len(results['businesses'])
# Use math.ceil to round up for the total number of pages of results.
n_pages = math.ceil((results['total']-n_results)/ results_per_page)
n_pages

[i] Data/results_in_progress_NO_vegan.json not found. Saving empty list to new file.
- 0 previous results found.


9

In [51]:
for i in tqdm_notebook( range(1,n_pages+1)):
    
    ## Read in results in progress file and check the length
    with open(JSON_FILE, 'r') as f:
        previous_results = json.load(f)
    ## save number of results for to use as offset
    n_results = len(previous_results)
    
    if (n_results + results_per_page) > 1000:
        print('Exceeded 1000 api calls. Stopping loop.')
        break
    
    ## use n_results as the OFFSET 
    results = yelp_api.search_query(location=location,
                                    term=term, 
                                    offset=n_results)
    
    
    
    ## append new results and save to file
    previous_results.extend(results['businesses'])
    
    # display(previous_results)
    with open(JSON_FILE,'w') as f:
        json.dump(previous_results,f)
    
    time.sleep(.2)

  0%|          | 0/9 [00:00<?, ?it/s]

# After the loop has finished

Convert .json to dataframe
Load in the "results in progress" JSON file into a DataFrame:

In [52]:
# load final results
final_df = pd.read_json(JSON_FILE)
display(final_df.head(), final_df.tail())

id                               alias  \
0  VXVrYjLRqNpQY-ogN9ztmQ            cleveland-vegan-lakewood   
1  sS37silzgDcQxarNrJk2WQ             thyme-table-bay-village   
2  r4dghMWRoFmBBFlAQvpwrA         cockys-bagels-north-olmsted   
3  EAiFR299WV3OQ68pHMlmGw          phnom-penh-north-olmsted-2   
4  yBbqFPuwFiaom435KOY4cg  mashiso-asian-grille-north-olmsted   

                   name                                          image_url  \
0       Cleveland Vegan  https://s3-media4.fl.yelpcdn.com/bphoto/-Do3Co...   
1           Thyme Table  https://s3-media1.fl.yelpcdn.com/bphoto/p0AbvB...   
2        Cocky's Bagels  https://s3-media1.fl.yelpcdn.com/bphoto/f3IcYM...   
3            Phnom Penh  https://s3-media3.fl.yelpcdn.com/bphoto/k6nddW...   
4  Mashiso Asian Grille  https://s3-media2.fl.yelpcdn.com/bphoto/Bepp8m...   

   is_closed                                                url  review_count  \
0      False  https://www.yelp.com/biz/cleveland-vegan-lakew...           295   
1      False  https://www.yelp.com/biz/thyme-table-bay-villa...           172   
2      False  https://www.yelp.com/biz/cockys-bagels-north-o...           194   
3      False  https://www.yelp.com/biz/phnom-penh-north-olms...           118   
4      False  https://www.yelp.com/biz/mashiso-asian-grille-...           152   

                                          categories  rating  \
0  [{'alias': 'vegan', 'title': 'Vegan'}, {'alias...     4.5   
1  [{'alias': 'newamerican', 'title': 'American (...     4.5   
2  [{'alias': 'bagels', 'title': 'Bagels'}, {'ali...     4.5   
3  [{'alias': 'cambodian', 'title': 'Cambodian'},...     4.0   
4  [{'alias': 'asianfusion', 'title': 'Asian Fusi...     4.0   

                                         coordinates        transactions  \
0  {'latitude': 41.4855267595181, 'longitude': -8...          [delivery]   
1  {'latitude': 41.479158, 'longitude': -81.9207637}  [pickup, delivery]   
2   {'latitude': 41.419271, 'longitude': -81.920495}  [pickup, delivery]   
3     {'latitude': 41.41678, 'longitude': -81.92207}          [delivery]   
4   {'latitude': 41.419502, 'longitude': -81.898061}  [pickup, delivery]   

  price                                           location         phone  \
0    $$  {'address1': '17112 Detroit Ave', 'address2': ...  +12162210201   
1    $$  {'address1': '583 Dover Center Rd', 'address2'...  +14406176964   
2     $  {'address1': '26703 Brookpark Extension', 'add...  +14404551205   
3    $$  {'address1': '27080 Lorain Rd', 'address2': ''...  +12162510210   
4     $  {'address1': '24954 Brookpark Rd', 'address2':...  +14408011184   

    display_phone      distance  
0  (216) 221-0201  10908.787939  
1  (440) 617-6964   6628.324334  
2  (440) 455-1205    596.588885  
3  (216) 251-0210    797.602452  
4  (440) 801-1184   1277.448497

id                                alias  \
173  iGJFu5WofKIGD9nRj1-zRQ  heinens-grocery-store-rocky-river-2   
174  a1DFzgQRYAmfA5o1fDVPEQ            insomnia-cookies-lakewood   
175  f0ELDvCsvHaWwxGodOd8Sw              inca-tea-cafe-cleveland   
176  nUogaVQovKvmKHDFGmFlQA        starbucks-coffee-cleveland-19   
177  MUMiGoEZ3NBwebH4Eff4LA            dq-orange-julius-lakewood   

                       name  \
173  Heinen's Grocery Store   
174        Insomnia Cookies   
175           Inca Tea Cafe   
176        STARBUCKS COFFEE   
177      DQ / Orange Julius   

                                             image_url  is_closed  \
173  https://s3-media2.fl.yelpcdn.com/bphoto/f4bpyS...      False   
174  https://s3-media1.fl.yelpcdn.com/bphoto/NdonkA...      False   
175  https://s3-media1.fl.yelpcdn.com/bphoto/JQ8OYY...      False   
176  https://s3-media1.fl.yelpcdn.com/bphoto/PkiRXF...      False   
177  https://s3-media3.fl.yelpcdn.com/bphoto/XmLCXd...      False   

                                                   url  review_count  \
173  https://www.yelp.com/biz/heinens-grocery-store...            78   
174  https://www.yelp.com/biz/insomnia-cookies-lake...            53   
175  https://www.yelp.com/biz/inca-tea-cafe-clevela...            41   
176  https://www.yelp.com/biz/starbucks-coffee-clev...            28   
177  https://www.yelp.com/biz/dq-orange-julius-lake...            22   

                                            categories  rating  \
173         [{'alias': 'grocery', 'title': 'Grocery'}]     4.0   
174  [{'alias': 'desserts', 'title': 'Desserts'}, {...     3.0   
175  [{'alias': 'coffee', 'title': 'Coffee & Tea'},...     4.0   
176     [{'alias': 'coffee', 'title': 'Coffee & Tea'}]     3.5   
177  [{'alias': 'icecream', 'title': 'Ice Cream & F...     3.0   

                                           coordinates        transactions  \
173  {'latitude': 41.4818115, 'longitude': -81.8347...                  []   
174   {'latitude': 41.482078, 'longitude': -81.827128}  [delivery, pickup]   
175  {'latitude': 41.40854970720977, 'longitude': -...                  []   
176  {'latitude': 41.3715799848749, 'longitude': -8...          [delivery]   
177   {'latitude': 41.485052, 'longitude': -81.813922}  [delivery, pickup]   

    price                                           location         phone  \
173    $$  {'address1': '19219 Detroit Rd', 'address2': '...  +14403313830   
174     $  {'address1': '18508 Detroit Ave', 'address2': ...  +12165866313   
175     $  {'address1': '5300 Riverside Dr', 'address2': ...  +18444622832   
176    $$  {'address1': '17999 Bagley Rd', 'address2': ''...  +12169048217   
177     $  {'address1': '16803 Detroit Rd', 'address2': '...  +12165217763   

      display_phone      distance  
173  (440) 331-3830  10288.126588  
174  (216) 586-6313  10783.873583  
175  (844) 462-2832   6874.011333  
176  (216) 904-8217   9430.573324  
177  (216) 521-7763  10440.048359

Because our Yelp results include columns that contain lists, we cannot check every column in the dataframe for duplicates.
Instead, we can use the subset argument for df.duplicated() and df.drop_duplicates() to only check the id column for duplicates.s 

In [54]:
# check for duplicate ID's 
final_df.duplicated(subset='id').sum()

21

In [55]:
## Drop duplicate ids and confirm there are no more duplicates
final_df = final_df.drop_duplicates(subset='id')
final_df.duplicated(subset='id').sum()

0

Save the final DataFrame to a .csv (or a .csv.gz if it's too big for the GitHub file size limit).

In [56]:
# save the final results to a compressed csv
final_df.to_csv('Data/final_results_in_progress_NO_vegan.csv.gz', compression='gzip',index=False)